packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -54,27 +55,75 @@ export async function tryHotswapDeployment(
 }
 
-async function findAllHotswappableChanges(
-  stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
-): Promise<HotswapOperation[] | undefined> {
-  const promises = new Array<Promise<ChangeHotswapResult>>();
-  stackChanges.resources.forEachDifference(async (logicalId: string, change: cfn_diff.ResourceDifference) => {
-    promises.push(isHotswappableLambdaFunctionChange(logicalId, change, evaluateCfnTemplate));
-  });
-  return Promise.all(promises).then(hotswapDetectionResults => {
-    const hotswappableResources = new Array<HotswapOperation>();
-    let foundNonHotswappableChange = false;
-    for (const lambdaFunctionShortCircuitChange of hotswapDetectionResults) {
-      if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
-        foundNonHotswappableChange = true;
-      } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
-        // empty 'if' just for flow-aware typing to kick in...
-      } else {
-        hotswappableResources.push(lambdaFunctionShortCircuitChange);
-      }
+async function findAllHotswappableChanges(stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate):
+Promise<HotswapOperation[] | undefined> {
+  const hotswappableResources = new Array<HotswapOperation>();
+  let foundNonHotswappableChange = false;
+  const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
+
+  // gather the results of the detector functions
+  stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isResourceChangeHotswappable(change);
+
+    if (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+      foundNonHotswappableChange = true;
+    } else if (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT) {
+      // empty 'if' just for flow-aware typing to kick in...
+    } else {
+      const detectorResults = new Array<Promise<ChangeHotswapResult>>();
+      detectorResults.push(isHotswappableLambdaFunctionChange(logicalId, nonHotswappableResourceFound, evaluateCfnTemplate));
+      detectorResults.push(isHotswappableStateMachineChange(logicalId, nonHotswappableResourceFound, evaluateCfnTemplate));
+
+      promises.push(detectorResults);
     }
-    return foundNonHotswappableChange ? undefined : hotswappableResources;
   });
+
+  // resolve all detector results
+  for (const detectorResultPromises of promises) {
+    await Promise.all(detectorResultPromises).then(hotswapDetectionResults => {
+      for (const result of hotswapDetectionResults) {
+        if (typeof result !== 'string') {
+          hotswappableResources.push(result);
+        }
+      }
+
+      // if we found any hotswappable changes, return now
+      if (hotswappableResources.length > 0) {
```

Comment:
This doesn't make sense really.

`hotswappableResources` is the global table for these results (it's for all changes). While what we should do is have a per-change table of these, and only if that table is non-empty, should we short-circuit here.

Otherwise, if the first change is hotswappable, but the other 20 are not, you will still attempt to do a hotswap deployment, which is the wrong thing to do.

Please, first add a unit test showing this behavior (the unit test should fail), and then fix it here.